vendors/nordic/boards/nrf52840-dk/ports/common_io/iot_flash.c

```diff
@@ -0,0 +1,490 @@
+                IotLogError( "%s: Insufficient memory to queue operation. Review sdk_config.h:FDS_OP_QUEUE_SIZE", __func__ );
+                lReturnCode = IOT_FLASH_ERASE_FAILED;
+                break;
+
+            case NRF_ERROR_INVALID_ADDR:
+                IotLogError( "%s: Out-of-bounds access (0x%x)", __func__, ulAddress );
+                lReturnCode = IOT_FLASH_INVALID_VALUE;
+                break;
+
+            default:
+                lReturnCode = IOT_FLASH_INVALID_VALUE;
+                break;
+        }
+    }
+
+    return lReturnCode;
+}
+int32_t iot_flash_read_sync( IotFlashHandle_t const pxFlashHandle,
+                             uint32_t ulAddress,
+                             uint8_t * const pvBuffer,
+                             size_t xBytes )
+{
+    int32_t lReturnCode = IOT_FLASH_INVALID_VALUE;
+
+    if( !prvIsOperableHandle( pxFlashHandle ) )
+    {
+        lReturnCode = IOT_FLASH_INVALID_VALUE;
+    }
+    else if( nrf_fstorage_is_busy( &iot_fstorage ) )
+    {
+        lReturnCode = IOT_FLASH_DEVICE_BUSY;
+    }
+    else
+    {
+        ret_code_t xReturnCode_NRF = nrf_fstorage_read( &iot_fstorage, ulAddress, pvBuffer, xBytes );
+
+        switch( xReturnCode_NRF )
+        {
+            case NRF_SUCCESS:
+                lReturnCode = IOT_FLASH_SUCCESS;
+                break;
+
+            case NRF_ERROR_NO_MEM:
+                IotLogError( "%s: Insufficient memory to queue operation. Review sdk_config.h:FDS_OP_QUEUE_SIZE", __func__ );
+                lReturnCode = IOT_FLASH_ERASE_FAILED;
+                break;
+
+            case NRF_ERROR_INVALID_ADDR:
+                IotLogError( "%s: Out-of-bounds access (0x%x)", __func__, ulAddress );
+                lReturnCode = IOT_FLASH_INVALID_VALUE;
+                break;
+
+            default:
+                lReturnCode = IOT_FLASH_INVALID_VALUE;
+                break;
+        }
+    }
+
+    return lReturnCode;
+}
+
+/* If a soft-device is present nrf_fstorage_sd is used, else nrf_fstorage_nvmc is used.
+ * While nrf_fstorage_sd does provide proper async functionality, nrf_fstorage_nvmc does not.
+ * See nrf_fstorage_nvmc:write(...)
+ */
+int32_t iot_flash_write_async( IotFlashHandle_t const pxFlashHandle,
+                               uint32_t ulAddress,
+                               uint8_t * const pvBuffer,
+                               size_t xBytes )
+{
+    return IOT_FLASH_FUNCTION_NOT_SUPPORTED;
+}
+
+int32_t iot_flash_read_async( IotFlashHandle_t const pxFlashHandle,
+                              uint32_t ulAddress,
+                              uint8_t * const pvBuffer,
+                              size_t xBytes )
+{
+    return IOT_FLASH_FUNCTION_NOT_SUPPORTED;
+}
+
+int32_t iot_flash_close( IotFlashHandle_t const pxFlashHandle )
+{
+    int32_t lReturnCode = IOT_FLASH_INVALID_VALUE;
+
+    if( prvIsOperableHandle( pxFlashHandle ) )
+    {
+        if( NRF_SUCCESS == nrf_fstorage_uninit( &iot_fstorage, NULL ) )
```

Comment:
This is a nit but it can be nice to separate the function result into a variable outside of the if statement to make debugging easier

Reply:
That's reasonable. Added in latest commit